action/protocol/generic_validator.go

```diff
@@ -66,69 +66,90 @@ func (v *GenericValidator) Validate(ctx context.Context, selp *action.SealedEnve
+	featureCtx, ok := GetFeatureCtx(ctx)
+	if ok && !featureCtx.EnableAccessListTx && selp.Version() == action.AccessListTxType {
+		return errors.Wrap(action.ErrInvalidAct, "access list tx is not enabled")
+	}
+	if ok && !featureCtx.EnableDynamicFeeTx && selp.Version() == action.DynamicFeeTxType {
+		return errors.Wrap(action.ErrInvalidAct, "dynamic fee tx is not enabled")
+	}
+	if ok && !featureCtx.EnableBlobTransaction && selp.Version() == action.BlobTxType {
+		return errors.Wrap(action.ErrInvalidAct, "blob tx is not enabled")
+	}
+	if ok && featureCtx.EnableDynamicFeeTx {
+		// check transaction's max fee can cover base fee
+		if selp.Envelope.GasFeeCap().Cmp(new(big.Int).SetUint64(action.InitialBaseFee)) < 0 {
+			return errors.Errorf("transaction cannot cover base fee, max fee = %s, base fee = %d",
+				selp.Envelope.GasFeeCap().String(), action.InitialBaseFee)
+		}
+		if selp.Envelope.GasTipCap().Cmp(MinTipCap) < 0 {
+			return errors.Wrapf(action.ErrUnderpriced, "tip cap is too low: %s, min tip cap: %s", selp.Envelope.GasTipCap().String(), MinTipCap.String())
+		}
+	}
+	if ok && featureCtx.EnableBlobTransaction && len(selp.BlobHashes()) > 0 {
+		// validate sidecar
+		if blkCtx, ok := GetBlockCtx(ctx); (ok && !blkCtx.SkipSidecarValidation) || selp.BlobTxSidecar() != nil {
+			if err := selp.ValidateSidecar(); err != nil {
+				return errors.Wrap(err, "failed to validate blob sidecar")
 			}
 		}
-		if ok && !featureCtx.EnableAccessListTx && selp.Version() == action.AccessListTxType {
-			return errors.Wrap(action.ErrInvalidAct, "access list tx is not enabled")
+	}
+	return v.validateWithState(ctx, selp)
+}
+
+func (v *GenericValidator) ValidateWithState(ctx context.Context, selp *action.SealedEnvelope) error {
+	return v.validateWithState(ctx, selp)
+}
+
+func (v *GenericValidator) validateWithState(ctx context.Context, selp *action.SealedEnvelope) error {
+	if action.IsSystemAction(selp) {
+		return nil
+	}
+	var (
+		caller         = selp.SenderAddress()
+		nonce          uint64
+		featureCtx, ok = GetFeatureCtx(ctx)
+	)
+	if caller == nil {
+		return errors.New("failed to get address")
+	}
+	if featureCtx.FixGasAndNonceUpdate || selp.Nonce() != 0 {
+		confirmedState, err := v.accountState(ctx, v.sr, caller)
+		if err != nil {
+			return errors.Wrapf(err, "invalid state of account %s", caller.String())
+		}
+		if featureCtx.UseZeroNonceForFreshAccount {
+			nonce = confirmedState.PendingNonceConsideringFreshAccount()
+		} else {
+			nonce = confirmedState.PendingNonce()
 		}
-		if ok && !featureCtx.EnableDynamicFeeTx && selp.Version() == action.DynamicFeeTxType {
-			return errors.Wrap(action.ErrInvalidAct, "dynamic fee tx is not enabled")
+		if nonce > selp.Nonce() {
+			return action.ErrNonceTooLow
 		}
-		if ok && !featureCtx.EnableBlobTransaction && selp.Version() == action.BlobTxType {
-			return errors.Wrap(action.ErrInvalidAct, "blob tx is not enabled")
+	}
+	if !ok {
+		return nil
+	}
+	if featureCtx.SufficentBalanceGuarantee {
```

Comment:
basefee validation of dynamicFeeTx also should be moved, thus next PR will change to use basefee of block context instead of the fixed initialBaseFee